### ISSUE TITLE
Improve dashboards for customized source labels

### ISSUE THREAD

**Issue by jotak** [kiali/kiali#2637]
Today, the custom (runtimes) dashboards system requires to have `app` and `version` labels set on metrics, to identify the source of the metrics. This is basically assuming that metrics are produced from a Kubernetes service discovery configured in Prometheus, with label-mapping turned on.

In the production-scale prometheus scenario ( see https://github.com/kiali/kiali/issues/2518 ), this assumption is wrong and the labels slightly differ.

The goal of this task would be to add new fields in dashboards CR to override source identifying labels per dashboard basis.

(Note that it can be worked-around by relabelling, but it still would be better to have Kiali being more flexible on that).

**Comment by jshaughn** [comment]
I'm not sure if it extends properly but app/version labels are deprecated in Istio metrics and replaced with canonical_service and canonical_revision.

**Comment by jotak** [comment]
@jshaughn yes, which makes it even more interesting to have it flexible / defined in dashboard CR

**Comment by lucasponce** [comment]
I'm revisiting this issue.
Initially it sounds interesting, but when I reviewed the code, I think it can create more implications, so I've changed a little bit my thought about this.

Today, custom dashboards are linked per "App" and "Workload" assuming that metrics have "app" / "version" labels on them.

The initial query is triggered from the UI, and when "app" or "version" is changed globally, it's reflected:

```
  private initOptions(settings: MetricsSettings): DashboardQuery {
    const filters = `${serverConfig.istioLabels.appLabelName}:${this.props.app}`;
    const options: DashboardQuery = this.props.version
      ? {
          labelsFilters: `${filters},${serverConfig.istioLabels.versionLabelName}:${this.props.version}`
        }
      : {
          labelsFilters: filters,
          additionalLabels: 'version:Version'
        };
    MetricsHelper.settingsToOptions(settings, options, []);
    return options;
  }
```

From implementation side, try to personalize this is like asking "ok, other metrics may have a different way to link with a proper app/workload" and then for Istio metrics there will be used some labels but for custom metrics there should be another type of filtering.

Then I think it wouldn't be as simple as to change the labels in the filter, but to provide some kind of expression language where a filter can take a different label AND linked with a different value (think like "my-new-label" should be equal to $workloadName or another variable that should be handled via a rich expression).

So, then that would add some level of complexity hard to maintain IMO.

As we are moving to use the custom dashboard to personalize Envoy metrics, perhaps it's good to keep that requeriment (that metrics should keep using same app/version labels used in Istio - again, user can change them, but for the all context -).

Also, this will trigger some changes in the current implementation as this expressions should be added in the workload as an extension of the dashboard references.

So, I'd vote to close this, keep it in the radar as it may be revisit in the future.
